### ISSUE TITLE
Modeling for database samples and tokenomics

### ISSUE THREAD

**Issue by tjayrush** [Great-Hill-Corporation/quickBlocks#520]
This is a first pass, but I think I can do pretty much everything I want to do with this. Is this what you were looking for? Does it make sense?

- A monitor is a list of Ethereum addresses
- Each address can participate as either the ‘to’ or the ‘from’ address in a transaction.
- A transaction either sends ether from one account to another or calls a function on one account from another
- A block’s timestamp is the smallest granularity of time (usually 14 seconds), therefor all transactions a block have the same timestamp and as a result the same price at time of transaction
- I used hashes as the relationship keys. Hashes are verbose and probably take up too much space, but they are perfect unique identifiers and therefore keys.
- I’m not sure I have the modeling of the ‘to’ and ‘from’ correct, but I want to be able to query for all transactions from, all transactions to, and all transaction either in or out of an account.

<img width="1751" alt="screen shot 2017-11-10 at 12 01 02 am" src="https://user-images.githubusercontent.com/5417918/32643596-6be808e0-c5aa-11e7-9c82-6e4117b9bec4.png">

I downloaded and used [mySQLWorkbench](https://www.mysql.com/products/workbench/) which is very nice and easy to use. I can share the file, if you need it. Let me know.

**Comment by crodnun** [comment]
Hi Thomas,
Here you have my comments about this diagram

**COMMENTS**

1) I think that the arrows that connect the tables are not correct, but we can ignore them for the moment. The arrows define the relationships and say how many entries in one table are connected with a given number of entries in the peer table. Anyways we can deal with this later.

2) A good practice is to add always a VARCHAR(N) field to identify those objects that can be named:

    Example: At monitor table we can add just a new field to say "Monitor 0 name is XXXXXX" this way we make the database human readable at the same time that we keep the ID to speed up the search.

The same happens with monitored accounts, we could add there a name for them (each account), this way you do not have to deal with numbers and remember which address goes with a given entity.

3) The 3 tables: Transactions, ToTransactions and FromTransactions seem redundant for me, we could  join them in  the same table, just adding the from and to fields to the Transactions table itself. We can obtain this information easily with a query ("Get every transaction to" or "Get every transaction from"), no need to create new tables for just classifying the transactions. The queries can do this for you.

If we do these small tweaks, I see that we will have two types of tables at this database:

1) Static tables: Monitors configuration (Monitor + MonitoredAccounts)
2) Dynamic tables: When running a monitor, it populates them with the information requested at static tables. Here we only store those blocks and transactions that contain valid information for the monitor.

Now how I see the logic:

**LOGIC**

Using them, we could implement this logic (except the database insertions, I think that probably is the logic that already use the monitors):

1) A monitor scrapes the blockchain and looks (at each block) for transactions where its addresses are involved.

2) If at current block, a matching is found. This block is inserted into database, and so do the transactions where the configured addreses are present (at from, to, or both).

3) This way you have populated the dynamic information for this monitor: blocks and transactions. And inside the transactions, you can store whole information and then just running queries to obtain the information you want.



Let me know your thoughts and we can implement this first database as an start. I can create it at stage server, fill with dummy data (just using mysql statements, no need to code them) to give something to be displayed online, and then we can move to connect the monitor with the database.

**Comment by tjayrush** [comment]
Can you query for (both from and to) at the same time?

I will take another stab at the design (just to practice). I understand what you mean about the arrows. I noticed that, but I forgot to do it.  I will post another design and then you can make comments on that, and then we can go forward with fake data. In the meantime, I can write the "real data" to a database on my local machine. I'll try to do it tonight.

I don't want to ask you to get on a call this weekend, but I will if you're available. Otherwise, let's do a call on Monday. I'm speaking to Vidal at 10:30 my time. Would it be possible for me to speak to you before Vidal on Monday? I get up early, so timing is no problem. Let me know.

**Comment by crodnun** [comment]
Hi Thomas, I have to travel this weekend, so better to meet on Monday. My availability on Monday is the following (using your local time):
From 8 am  to 9 am
From 10 am to 11 am.
From 2 pm to 6 pm.
Let me know if any of these slots works for you.

**Comment by tjayrush** [comment]
Let's talk at 8:30 am my time. I'll send a calendar invite.

**Comment by tjayrush** [comment]
Here's an updated model for the database. By the way, I used mySQL Workbench to make these images. I emailed you a copy of the file I used.

<img width="1608" alt="screen shot 2017-11-10 at 11 00 11 pm" src="https://user-images.githubusercontent.com/5417918/32686043-90d99472-c66b-11e7-828b-12f045de54a1.png">

**Comment by tjayrush** [comment]
Name of database: qb_monitors

**Comment by tjayrush** [comment]
For the Functions table, please add a VARCHAR(45) called FunctionGroup

**Comment by tjayrush** [comment]
This is the most up-to-date version of the database:

<img width="1436" alt="screen shot 2017-11-13 at 3 42 27 pm" src="https://user-images.githubusercontent.com/5417918/32748150-5f90e4ea-c889-11e7-9d0b-bcd9feba29e8.png">

**Comment by crodnun** [comment]
Hi Thomas, I have created the initial fake database (before your last update :-)), at staging server. 

You can visit it using your current remote connection (its name is qb_static_monitors). 

I attach here just an example of how you can create tables and databases using files and also how you can insert entries into the different tables (again using files). 

As you will learn, it is easy once you define the file to execute statements, redo, drop, create again, insert many entries. 

If you can, we can meet again tomorrow (I am free from 9AM onwards) to discuss the same, I think that I will need your help generating fake data (Mine probably is not as accurate as we would like) and also need to review some data (the addresses as you depicted here can not be BIGINT).

**Comment by tjayrush** [comment]
You are so good at what you do. I can't tell you how much I appreciate your work. You do exactly the right thing at the right time. (I'm struggling with the SQL syntax).

I'll go through this tonight and tomorrow morning before we speak at 9:00 am. Thanks.

**Comment by tjayrush** [comment]
I modified your wonderful script. Here it is:

[Database.txt](https://github.com/Great-Hill-Corporation/quickBlocks/files/1469065/Database.txt)

It produces this database:

<img width="1601" alt="screen shot 2017-11-13 at 6 55 03 pm" src="https://user-images.githubusercontent.com/5417918/32755700-8c867bac-c8a4-11e7-8cd1-370ff5f7f89c.png">

except for one thing. I can't figure out how to tie the three values blockNumber, transID, and traceID between TransactionsRef and Transactions table. 

The reason I want to do this is because I want to store the transactions only once. Every transaction has two sides -- from and to -- I may be misunderstanding, but it feels like if I store it the way we had it before I'll be storing more data, but we can discuss. If the previous version does not store extra data, then we can return to that, but I did change some names and field sizes, so we should start from here.

I'm going to work now storing some data into the databases.
